### PR TITLE
feat: implement API request for user-signup

### PR DESCRIPTION
**Pull request by teddy-f-47**
What's changed:
- ui/.env.development : changed vue's api base url to "http://localhost:8000/api" for now.
- ui/.env.production : changed vue's api base url to "http://localhost:8000/api" for now.
- ui/src/api/index.js : implemented API request function userSignup() using axios.
- ui/src/views/user/signup/SignupView.vue : now calls API's userSignup() on form submit, added handler upon "user already exists" and upon error.

Notable:
- It seems to be important to run the docker services in separate terminals/cmds. In my case, I need to have a separate terminal for each of `docker-compose up db`, `docker-compose up phpmyadmin`, `docker-compose up api`, `docker-compose up ui` and `docker-compose up nginx`. I can't 100% confirm but apparently running only one terminal with `docker-compose up` for all services is causing CORS issue with axios (but not with traditional fetch).
- There is no 'elegant' way to distinguish between "user was created successfully" and "user already exists", because both cases return success code 200 with no 'status' field. There is only 'message' field in the response. That is why I simply use `if (response["message"].includes(" already exists."))` to handle the case when user already exists.

Received email from backend:
```
Your confirmation link is here: http://localhost:8000/api/authentication/activate/InF5dmloeUBlbWEtc29maWEuZXUi.YqdUfQ.wbFdqcet7Mm7dylFs3QNSkpw6IQ
```

If user already exists:
user-signup1

Upon clicking account activation link:
user-signup2

### CODE REVIEW

File: ui/src/api/index.js [hunk 1]
```diff
@@ -38,3 +38,16 @@ export const getSurveys = () => {
 export const addSurvey = (data) => {
   return axios.post(ADD_SURVEY_URL, data);
 };
+
+const USER_SIGNUP_URL = "/authentication/signup";
+
+export const userSignup = (data) => {
+  let config = {
```
Comment (artikandri) [inline ui/src/api/index.js]: This approach is good enough if we only had one or two APIs; however we have a lot of APIs and so I think it's better to make use of the default configuration... I don't test this code so it may not work, but it should be something like this:
```
axios.defaults.headers.common['Accept'] = 'application/json'; 
axios.defaults.headers.common['Content-Type'] = 'application/json'; 
```
Can you check to make sure whether the above settings work or not?

File: ui/src/views/user/signup/SignupView.vue [hunk 2]
```diff
@@ -112,8 +114,43 @@ export default {
...
+          if (response["message"].includes(" already exists.")) {
+            this.$notify.toast(response["message"]);
+          }
+          else {
+            this.$router
+              .push({ name: "general-user-signup-thankyou" })
+              .catch(() => {});
+          }
+        })
+        .catch((error) => {
```
Comment (artikandri) [inline ui/src/views/user/signup/SignupView.vue:128]: Rather than putting the error message here, we can make use of the interceptor XD    I think I have put a basic interceptor there. It will be great if you can move the notification there, as well as  importing vuetify-notify there... but I dont remember if vuetify-notify can be exported or no XD   the package was as minimal as it can be. 

But if it's too much work, then this is OK.

File: ui/src/views/user/signup/SignupView.vue [hunk 3]
```diff
@@ -112,8 +114,43 @@ export default {
...
+          this.$notify.toast("Something went wrong. Please try again later.");
+          console.log(error);
+        });
+
+      /* WORKING CODE WITH FETCH
```
Comment (artikandri) [inline ui/src/views/user/signup/SignupView.vue]: Please remove unused code XD

**Review by artikandri — COMMENTED** [review]
Some minor things, but otherwise OK. Good job!

File: ui/src/api/index.js (hunk 1 above)
Comment (teddy-f-47) [reply ui/src/api/index.js]: This works. Will push my code soon.

File: ui/src/views/user/signup/SignupView.vue (hunk 2 above)
Comment (teddy-f-47) [reply ui/src/views/user/signup/SignupView.vue:128]: Yes, there is a basic interceptor, but I'm not sure how to import vuetify-notify into ui/src/api/index.js correctly. I suppose simply putting `this.$notify.toast("Something went wrong. Please try again later.");` inside the interceptor without importing vuetify-notify won't do it. And I'm afraid testing different ways to do that will spend way too much time (especially because re-building the ui service after every code change is really slow).

Therefore, I am leaving the error message as it is now, so that I can start working on the other API calls (I intend to work on user login next).

File: ui/src/api/index.js (hunk 1 above)
Comment (artikandri) [reply ui/src/api/index.js]: Great! When you worked on the login api, please also set the authentication in this way. I dont know how will the BE give the token, but when they did, please store it in vuex (if they havent stored already), so that we can use in the "Authentication" header XD but it is for later PR, not this one.

File: ui/src/views/user/signup/SignupView.vue (hunk 2 above)
Comment (artikandri) [reply ui/src/views/user/signup/SignupView.vue:128]: It wont do because vuetify-notify is part of the Vue environment, importing it will be a bit tricky XD so lets leave it for now.

File: ui/.env.development [hunk 4]
```diff
@@ -1,2 +1,2 @@
 NODE_ENV="development"
-VUE_APP_API_BASE_URL="localhost:8000/"
\ No newline at end of file
+VUE_APP_API_BASE_URL="http://localhost:8000/api"
```
Comment (artikandri) [inline ui/.env.development]: Sorry I just noticed, but I think its better to put the trailing slash here `/` rather than specifying it at the start of every single API url... how about it?

File: ui/.env.production [hunk 5]
```diff
@@ -1,2 +1,2 @@
 NODE_ENV="production"
-VUE_APP_API_BASE_URL="localhost:8000/"
\ No newline at end of file
+VUE_APP_API_BASE_URL="http://localhost:8000/api"
```
Comment (artikandri) [inline ui/.env.production]: same case

File: ui/src/api/index.js [hunk 6]
```diff
@@ -38,3 +41,9 @@ export const getSurveys = () => {
 export const addSurvey = (data) => {
   return axios.post(ADD_SURVEY_URL, data);
 };
+
+const USER_SIGNUP_URL = "/authentication/signup";
```
Comment (artikandri) [inline ui/src/api/index.js]: if you decided to add the trailing slash to the baseUrl, dont forget to remove the `/` from the start of the api url

File: ui/src/views/user/signup/SignupView.vue [hunk 7]
```diff
@@ -112,8 +114,21 @@ export default {
...
-      this.$router.push({ name: "general-user-signup-thankyou" });
+      userSignup(this.formData)
+        .then((response) => {
+          if (response["message"].includes(" already exists.")) {
+            this.$notify.toast(response["message"]);
+          }
+          else {
+            this.$router
+              .push({ name: "general-user-signup-thankyou" })
+              .catch(() => {});
```
Comment (artikandri) [inline ui/src/views/user/signup/SignupView.vue:125]: I noticed this in the SettingsView as well, but I forgot to comment back then. I think rather than having the catch block not doing anything, its better to remove it XD  Or do you have any reason on why did you add the block?

**Review by artikandri — COMMENTED** [review]
Minor things I missed previously, please check them and tell me if they are a bit too much XD

File: ui/src/views/user/signup/SignupView.vue (hunk 7 above)
Comment (teddy-f-47) [reply ui/src/views/user/signup/SignupView.vue:125]: Without the 'catch' block, router push throws an error into the console saying something about navigation guard. Well, it is not a dangerous error. Reference: [stackoverflow](https://stackoverflow.com/questions/62223195/vue-router-uncaught-in-promise-error-redirected-from-login-to-via-a)
The error won't appear on the ui, but I don't like seeing errors in the browser's console (and I imagine a user with IT know-how who happens to open their console while using the app will think something is wrong upon seeing the error). This is one way to 'eliminate' the error XD

Comment (artikandri) [reply ui/src/views/user/signup/SignupView.vue:125]: Aha XD  those errors then should be handled in router/index.js..  Alright, approved.

Comment (artikandri) [reply ui/src/views/user/signup/SignupView.vue:125]: Thank you for the good job, and congrats for your first axios and Vue plugin configuration XD

**Review by artikandri — APPROVED** [review]
